downstairs/src/lib.rs

```diff
@@ -833,40 +851,73 @@ where
                             bail!("expected version 1, got {}", version);
                         }
                         negotiated = 1;
-                        upstairs_uuid = Some(upstairs_id);
+                        upstairs_connection = Some(UpstairsConnection {
+                            upstairs_id,
+                            session_id,
+                            gen,
+                        });
                         println!("upstairs {:?} connected",
-                            upstairs_uuid.unwrap());
+                            upstairs_connection.unwrap());
 
                         let mut fw = fw.lock().await;
                         fw.send(Message::YesItsMe { version: 1 }).await?;
                     }
-                    Some(Message::PromoteToActive { upstairs_id }) => {
+                    Some(Message::PromoteToActive { upstairs_id, session_id, gen }) => {
                         if negotiated != 1 {
                             bail!("Received activate out of order {}",
                                 negotiated);
                         }
+
                         // Only allowed to promote or demote self
-                        if upstairs_uuid.unwrap() != upstairs_id {
+                        let mut upstairs_connection = upstairs_connection.unwrap();
+                        let matches_self =
+                            upstairs_connection.upstairs_id == upstairs_id &&
+                            upstairs_connection.session_id == session_id;
+
+                        if !matches_self {
                             let mut fw = fw.lock().await;
                             fw.send(
-                                Message::UuidMismatch { expected_id: upstairs_uuid.unwrap() }
+                                Message::UuidMismatch {
+                                    expected_id: upstairs_connection.upstairs_id,
+                                }
                             ).await?;
+
                             /*
                              * At this point, should we just return error?
                              * XXX
                              */
                         } else {
+                            // matches_self above should include a check for gen,
+                            // but gen number may change between negotiation and
+                            // activation (see `XXX because the generation number`
+                            // upstairs). update generation number here.
```

Comment:
I agree, this is currently required because of the upstairs `set_generation` call below the `XXX because the generation number is not plumbed up yet in` comment.

Reply:
How about we code up the check, and, oh.. you already did that.
Yeah, the code here is good.  The `println!("warning:` will become a `bail!` when we are ready for full generation number support.